File: Aeds/AedsII/Atividades/Trabalhos Praticos/Tp3/Num6/codigo.java
import java.io.BufferedReader;
import java.io.FileReader;


class Jogador {

    /*
     * Aluno(Dev): Tulio Gomes Braga;
     * Usuario: 1441272;
     * 
     * Descrição do conteudo:
     * 
     * Atributos da classe jogador, seguindo a tabela "players.csv";
     * Metodos sets e gets para manipular os valores do objeto;
     * Construtor: esse metodo fica responsavel por cria um objeto com valores nulos;
     * Clone: objeto que chama tem seus atributos copiados pra uma nova instância de objeto;
     * Inserir Elementos: Recebe valores ate ser igual a "FIM", cria um novo objeto, tenta transformar em inteiro e chama os proximos metodos;
     * SetaId: Recebe um id, abre a leitura no arquivo e vai percorrendo até o id lido seja igual ao recebido, caso for pega a linha correspondente;
     * SetaDados: Recebe uma linha, fragmenta de acordo com a virgula e insere os argumentos daquele objeto corrente;
     * MostraDados: Mostra todos os atributos do objeto corrente;
     * MostrarLista: Metodo pra percorrer toda nossa lista e printar os dados;
     * Escreve: Recebe o objeto corrente e printa o retorno de outro metodo;
     * toString: Similar ao mostrar dados, retorna os dados concatenados em String;
     * 
     * Exercicio: Manipulando o array a partir de criterios da pilha flexivel, no qual os elementos estao controlados por um ponteiro topo.
    */

    public static Jogador topo = null;
    public Jogador prox;

    private Integer id,altura,peso,anoNascimento;
    private String nome,universidade,cidadeNascimento,estadoNascimento;



    public String getEstadoNascimento() {
        return estadoNascimento;
    }
    public String getCidadeNascimento() {
        return cidadeNascimento;
    }
    public String getUniversidade() {
        return universidade;
    }
    public String getNome() {
        return nome;
    }
    public int getAnoNascimento() {
        return anoNascimento;
    }
    public int getPeso() {
        return peso;
    }
    public int getAltura() {
        return altura;
    }
    public int getId() {
        return id;
    }
    public void setId(int id) {
        this.id = id;
    }
    public void setAltura(int altura) {
        this.altura = altura;
    }
    public void setPeso(int peso) {
        this.peso = peso;
    }
    public void setAnoNascimento(int anoNascimento) {
        this.anoNascimento = anoNascimento;
    }
    public void setNome(String nome) {
        this.nome = nome;
    }
    public void setUniversidade(String universidade) {
        this.universidade = universidade;
    }
    public void setCidadeNascimento(String cidadeNascimento) {
        this.cidadeNascimento = cidadeNascimento;
    }
    public void setEstadoNascimento(String estadoNascimento) {
        this.estadoNascimento = estadoNascimento;
    }

    public Jogador(){
        this.id = null;
		this.nome = null;
		this.altura = null;
		this.peso = null;
		this.universidade = null;
		this.anoNascimento = null;
		this.cidadeNascimento = null;
		this.estadoNascimento = null;
        this.prox = null;
    }

    public Jogador cloneJogador() {
		Jogador clone = new Jogador();
		clone.id = this.getId();
		clone.nome = this.getNome();
		clone.altura = this.getAltura();
		clone.peso = this.getPeso();
		clone.universidade = this.getUniversidade();
		clone.anoNascimento = this.getAnoNascimento();
		clone.cidadeNascimento = this.getCidadeNascimento();
		clone.estadoNascimento = this.getEstadoNascimento();
		return clone;
	}

    /*----------------------------------------------------------------------------------------------------------------------------------------- INSERCAO DE DADOS INICIO */

    public void inserirElementos(){
        String input = MyIO.readLine();
        while (!input.equals("FIM")) {
            try {
                Integer id = Integer.parseInt(input);
                Jogador jogador = new Jogador();
                jogador.SetaId(id);                           

                jogador.prox = topo;                    // LIGA NO ANTIGO TOPO
                topo = jogador;                         // NOVO ELEMENTO SE TORNA TOPO
                jogador = null;
            } catch (Exception e) {}
            input = MyIO.readLine();
        }
        
    }

    public void SetaId(int x) throws Exception {
        FileReader file = new FileReader("players.csv"); 
        BufferedReader buffer = new BufferedReader(file);
        String linha;

        while ((linha = buffer.readLine()) != null) {
            String numId = linha.substring(0, linha.indexOf(","));

            try{
                if(Integer.parseInt(numId) == x) {
                    this.setaDados(linha);
                    break;
                }
            }catch(NumberFormatException e){}
        }
        buffer.close();
        file.close();
    }

    public void setaDados(String linha){
        String[] parte = linha.split(",");

        this.setId(Integer.parseInt(parte[0]));
        this.setNome(parte[1]);
        this.setAltura(Integer.parseInt(parte[2]));
        this.setPeso(Integer.parseInt(parte[3]));
        if(parte.length>4 && !parte[4].isEmpty()) this.setUniversidade(parte[4]); else this.setUniversidade("nao informado");
        this.setAnoNascimento(Integer.parseInt(parte[5]));
        if (parte.length>6 && !parte[6].isEmpty()) this.setCidadeNascimento(parte[6]); else this.setCidadeNascimento("nao informado");
        if (parte.length>7 && !parte[7].isEmpty()) this.setEstadoNascimento(parte[7]); else this.setEstadoNascimento("nao informado");

    }

	public void mostraPilha(Jogador i) {
		if (i != null) {
			mostraPilha(i.prox);
			i.Escrever();
		}
	}

    public void Escrever() {
		String linha = this.toString();
		MyIO.println(linha);
	}

    public String toString() {
		String txt = "[" + "] " + "## ";
        txt += this.getId() + " ## ";
		txt += this.getNome() + " ## ";
		txt += this.getAltura() + " ## ";
		txt += this.getPeso() + " ## ";
		txt += this.getAnoNascimento() + " ## ";
		txt += this.getUniversidade() + " ## ";
		txt += this.getCidadeNascimento() +  " ## ";
		txt += this.getEstadoNascimento() + " ##";
		return txt;
	}

    /*------------------------------------------------------------------------------------------------------------------------------------------ FIM INSERCAO DE DADOS */
    /* ------------------------------------------------------------------------------------------------------------------------------------------- INICIO MANIPULACAO */

    public static void manipulaElementos(){
        int qtd = MyIO.readInt();

        for(int i=0;i<qtd;i++){
            String input = MyIO.readLine();
            decodificaInput(input);
        }
    }

    public static void decodificaInput(String input){
        /*
         * I empilhar/inserir
         * R remover 
         */
        String[] parte = input.split(" ");

        if(parte.length == 1){
            String entrada = parte[0];
            if(entrada.equals("R")) removerFim();

        }
        else if(parte.length == 2){
            String entrada = parte[0];
            if(entrada.equals("I")){
                int idJogador =  Integer.parseInt(parte[1]);
                empilhar(idJogador);
            } 
        }
    }

    public static void empilhar(int idJogador){
        try{
            Jogador jogador = new Jogador();
            jogador.SetaId(idJogador);                           

            jogador.prox = topo;                    // LIGA NO ANTIGO TOPO
            topo = jogador;                         // NOVO ELEMENTO SE TORNA TOPO
            jogador = null;
        }catch(Exception e){}
    } 

    public static void removerFim(){
        Jogador tmp = topo;
        topo = topo.prox;
        tmp.prox = null; tmp = null;
    }

    /* ------------------------------------------------------------------------------------------------------------------------------------------- FIM MANIPULACAO */
    /* ------------------------------------------------------------------------------------------------------------------------------------------- INICIO MAIN */


    public static void main(String[] args) {            // Compilar:    javac .\Jogador.java        Executar:    type .\pub.in | java Jogador > result.txt
        Jogador jogador = new Jogador();                // Compilar:    javac Jogador.java          Executar:    java Jogador < pub.in > result.txt             

        jogador.inserirElementos();
        manipulaElementos();
        //jogador.mostraPilha(topo);
    }
    
    /* ----------------------------------------------------------------------------------------------------------------------------------------------- FIM MAIN */
}

File: Aeds/AedsII/Atividades/Trabalhos Praticos/Tp3/Num10/codigo.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

/*
    Dev: Tulio Gomes Braga
    Matricula: 1441272
    Curso: CC-PucMinas - 15/11/2023
*/

typedef struct {
    int id;
    int altura;
    int peso;
    int anoNascimento;
    char nome[100];
    char universidade[100];
    char cidadeNascimento[100];
    char estadoNascimento[100];

} Jogador;

typedef struct Celula {
    Jogador elemento;
    struct Celula* prox; 

}Celula;
Celula* newCelula(Jogador elemento){
    Celula* nova = (Celula*) malloc(sizeof(Celula));
    nova->elemento = elemento;
    nova->prox = NULL;

    return nova;    
}


Celula* topo;
int tamanho = 0, mostrar = 0;


void inserirElementos();
void setaId(int id);
void parteDados(const char *linha);
void setaDados(char *linha);
void mostraDados(Jogador jogador);
void mostrarTodos();
void removerInicioPrint();
void manipulaElementos();
void decodifica(char *entrada); 


void inserirElementos(){
    char input[100];
    scanf("%s", input);

    while(strcmp(input, "FIM")!= 0){
        int id = atoi(input);
        setaId(id);
        scanf("%s", input);
    }
}

void setaId(int id) {
    FILE *file;
    file = fopen("/tmp/players.csv", "r");

    char linha[110];
    char resp[110];
    bool achou = false;
    char *token;

    fgets(linha, sizeof(linha), file);

    while (fgets(linha, sizeof(linha), file) != NULL && !achou){
        strcpy(resp,linha);
        token = strtok(linha,",");
        int num = atoi(token);

        if(id==num){
            parteDados(resp);
            achou=true;
        }
        token=strtok(NULL,",");

    }
    fclose(file);
}

void parteDados(const char *x) {

    int len = strlen(x);
    char linha[len];
    strcpy(linha,x);

    linha[len-1]='\0';
    int j = 0;
    char *nvlinha = malloc(len + 4);


    for (int i = 0; i < len; i++) {
        if (linha[i]==',' && linha[i+1]==',') {
            nvlinha[j]=linha[i];
            nvlinha[j+1]=' ';
            nvlinha[j+2]=',';
            j+=3;
            i++;
        } else {
            nvlinha[j] = linha[i];
            j++;
        }
    }

    if(nvlinha[j-2]==','){
        nvlinha[j-1]=' ';
    }
    nvlinha[j]='\0';

    setaDados(nvlinha);
    free(nvlinha);
}

void setaDados(char *linha){
    Jogador jogador;

    int t=0;
    char *parte[8];
    char *token = strtok(linha, ",");

    while (token != NULL && t < 8) {
        parte[t] = token;
        token = strtok(NULL, ",");          //MANTEM UM ESTADO INTERNO PARA RASTREAR A POSICAO ATUAL NA STRING
        t++;
    }

    jogador.id = atoi(parte[0]);
    if(strcmp(parte[1], " ")== 0)strcpy(jogador.nome, "nao informado");else strcpy(jogador.nome, parte[1]);
    jogador.altura = atoi(parte[2]);
    jogador.peso = atoi(parte[3]);
    if(strcmp(parte[4], " ")== 0)strcpy(jogador.universidade, "nao informado");else strcpy(jogador.universidade, parte[4]);
    jogador.anoNascimento = atoi(parte[5]);
    if(strcmp(parte[6], " ")== 0)strcpy(jogador.cidadeNascimento, "nao informado");else strcpy(jogador.cidadeNascimento, parte[6]);
    if(strcmp(parte[7], " ")== 0)strcpy(jogador.estadoNascimento, "nao informado");else strcpy(jogador.estadoNascimento, parte[7]);

    //mostraDados(jogador);

    Celula* tmp = newCelula(jogador);   
    tmp->prox = topo;   
    topo = tmp; tmp = NULL;
    tamanho++;


}

void mostraDados(Jogador jogador){
    printf("[%i] ## %s ## %d ## %d ## %d ## %s ## %s ## %s ##\n",
        mostrar,
        jogador.nome,   
        jogador.altura,
        jogador.peso,
        jogador.anoNascimento,
        jogador.universidade,
        jogador.cidadeNascimento,
        jogador.estadoNascimento    
    );
    mostrar++;
}

void mostrarTodos(){
    Celula* i;
    for(i=topo; i!=NULL; i=i->prox){
        mostraDados(i->elemento);
    }
}

void mostrarTodosR(Celula *i){
    if(i!=NULL){
        mostrarTodosR(i->prox);
        mostraDados(i->elemento);
    }
}

void removerInicioPrint(){
    printf("(R) %s\n", topo->elemento.nome);
    Celula* tmp = topo;
    topo = topo->prox; tamanho--;
    tmp->prox = NULL;
}

void manipulaElementos(){
    int qtd; scanf("%i",&qtd);
    char entrada[20];

    while (getchar() != '\n');

    for (int i=0;i<qtd;i++){
        fgets(entrada, sizeof(entrada), stdin);                     // O FGETS COLOCA UM "\N" NO FINAL DA STRING;
        decodifica(entrada);
    }
}

void decodifica(char *entrada){
    int len = strlen(entrada);
    if (len>0 && entrada[len-1]=='\n'){entrada[len-1]='\0';}
    char *token = strtok(entrada," ");

    if(token!=NULL){
        if (strcmp(token, "I")==0){
            token = strtok(NULL, " ");
            setaId(atoi(token));
        }
        else if(strcmp(entrada, "R")==0){
            removerInicioPrint();
        }
    }
}



int main(){ 
    inserirElementos();
    manipulaElementos();
    //mostrarTodos();
    mostrarTodosR(topo);
    return 0;
}